README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# funneljoin <img src='man/figures/logo.png' align="right" height="139" />

[![Travis-CI Build
Status](https://travis-ci.org/robinsones/funneljoin.svg?branch=master)](https://travis-ci.org/robinsones/funneljoin)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/funneljoin)](https://cran.r-project.org/package=funneljoin)

The goal of funneljoin is to make it easy to analyze behavior funnels.
For example, maybe you’re interested in finding the people who visit a
page and then register. Or you want all the times people click on an
item and add it to their cart within 2 days. These can all be answered
quickly with funneljoin’s `after_join()` or `funnel_start()` and
`funnel_step()`. As funneljoin uses dplyr, it can also work with remote
tables, **but has only been tried on postgres**.

For more examples of how to use funneljoin, check out [the
vignette](https://robinsones.github.io/funneljoin/articles/funneljoin.html),
which shows different types of joins and the optional arguments, or this
[blog
post](https://hookedondata.org/introducing-the-funneljoin-package/),
which showcases how to use funneljoin analyze questions and answers on
StackOverflow.

## Installation

You can install this package from CRAN:

``` r
install.packages("funneljoin")
```

Or you can install the development version from GitHub with
[remotes](https://github.com/r-lib/remotes):

``` r
library(remotes)
install_github("robinsones/funneljoin")
```

## after\_join()

``` r
library(dplyr)
library(funneljoin)
```

We’ll take a look at two tables that come with the package, `landed` and
`registered`. Each has a column `user_id` and `timestamp`.

Let’s say we wanted to get the first time people landed and the first
time afterward they registered. We would `after_inner_join()` with a
`first-firstafter` type:

``` r
landed %>%
  after_inner_join(registered, 
                   by_user = "user_id",
                   by_time = "timestamp",
                   type = "first-firstafter",
                   suffix = c("_landed", "_registered"))
#> # A tibble: 5 x 3
#>   user_id timestamp_landed timestamp_registered
#>     <dbl> <date>           <date>              
#> 1       1 2018-07-01       2018-07-02          
#> 2       4 2018-07-01       2018-07-02          
#> 3       3 2018-07-02       2018-07-02          
#> 4       6 2018-07-07       2018-07-10          
#> 5       5 2018-07-10       2018-07-11
```

The first two arguments are the tables we’re joining, with the first
table being the events that happen first. We then specify:

  - `by_time`: the time columns in each table. This would typically be a
    datetime or a date column. These columns are used to filter for time
    y being after or the same as time x.
  - `by_user`:the user or identity columns in each table. These must be
    identical for a pair of rows to match.
  - `type`: the type of funnel used to distinguish between event pairs,
    such as “first-first”, “last-first”, “any-firstafter”.
  - `suffix` (optional): just like dplyr’s join functions, this
    specifies what should be appended to the names of columns that are
    in both tables.

`type` can be any combination of `first`, `last`, `any`, and
`lastbefore` with `first`, `last`, `any`, and `firstafter`. Some common
ones you may use include:

  - **first-first**: Take the earliest x and y for each user **before**
    joining. For example, you want the first time someone entered an
    experiment, followed by the first time someone **ever** registered.
    If they registered, entered the experiment, and registered again,
    you do not want to include that person.
  - **first-firstafter**: Take the first x, then the first y after that.
    For example, you want when someone first entered an experiment and
    the first course they started afterwards. You don’t care if they
    started courses before entering the experiment.
  - **lastbefore-firstafter**: First x that’s followed by a y before the
    next x. For example, in last click paid ad attribution, you want the
    last ad someone clicked before the first subscription they did
    afterward.
  - **any-firstafter**: Take all Xs followed by the first Y after it.
    For example, you want all the times someone visited a homepage and
    their first product page they visited afterwards.
  - **any-any**: Take all Xs followed by all Ys. For example, you want
    all the times someone visited a homepage and **all** the product
    pages they saw afterward.

If your time and user columns have different names, you can work with
that too:

``` r
landed <- landed %>%
  rename(landed_at = timestamp,
         user_id_x = user_id)

registered <- registered %>%
  rename(registered_at = timestamp,
         user_id_y = user_id)
```

``` r
landed %>%
  after_inner_join(registered, 
                   by_user = c("user_id_x" = "user_id_y"),
                   by_time = c("landed_at" = "registered_at"),
                   type = "first-first")
#> # A tibble: 4 x 3
#>   user_id_x landed_at  registered_at
#>       <dbl> <date>     <date>       
#> 1         1 2018-07-01 2018-07-02   
#> 2         3 2018-07-02 2018-07-02   
#> 3         6 2018-07-07 2018-07-10   
#> 4         5 2018-07-10 2018-07-11
```

## funnel\_start() and funnel\_step()

Sometimes you have all the data you need in one table. For example,
let’s look at this table of user activity on a website.

``` r
activity <- tibble::tribble(
  ~ "user_id", ~ "event", ~ "timestamp",
  1, "landing", "2019-07-01",
  1, "registration", "2019-07-02",
  1, "purchase", "2019-07-07",
  1, "purchase", "2019-07-10",
  2, "landing", "2019-08-01",
  2, "registration", "2019-08-15",
  3, "landing", "2019-05-01",
  3, "registration", "2019-06-01",
  3, "purchase", "2019-06-04",
  4, "landing", "2019-06-13"
)
```

We can use `funnel_start()` and `funnel_step()` to make an activity
funnel. `funnel_start()` takes five arguments:

  - `tbl`: The table of events.
  - `moment_type`: The first moment, or event, in the funnel.
  - `moment`: The name of the column that indicates the `moment_type`.
  - `tstamp`: The name of the column with the timestamps of the moment.
  - `user`: The name of the column indicating the user who did the
    moment.

<!-- end list -->

``` r
activity %>%
  funnel_start(moment_type = "landing", 
               moment = "event", 
               tstamp = "timestamp", 
               user = "user_id")
#> # A tibble: 4 x 2
#>   user_id timestamp_landing
#>     <dbl> <chr>            
#> 1       1 2019-07-01       
#> 2       2 2019-08-01       
#> 3       3 2019-05-01       
#> 4       4 2019-06-13
```

`funnel_start()` returns a table with the user\_ids and a column with
the name of your timestamp column, `_`, and the moment type. This table
also includes metadata.

To add more moments to the funnel, you use `funnel_step()`. Since you’ve
indicated in `funnel_start()` what columns to use for each part, now you
only need to have the `moment_type` and the `type` of `after_join()`
(e.g. “first-first”, “first-any”).

``` r
activity %>%
  funnel_start(moment_type = "landing", 
               moment = "event", 
               tstamp = "timestamp", 
               user = "user_id") %>%
  funnel_step(moment_type = "registration",
              type = "first-firstafter")
#> # A tibble: 4 x 3
#>   user_id timestamp_landing timestamp_registration
#>     <dbl> <chr>             <chr>                 
#> 1       3 2019-05-01        2019-06-01            
#> 2       4 2019-06-13        <NA>                  
#> 3       1 2019-07-01        2019-07-02            
#> 4       2 2019-08-01        2019-08-15
```

You can continue stacking on `funnel_step()` with more moments.

``` r
activity %>%
  funnel_start(moment_type = "landing", 
               moment = "event", 
               tstamp = "timestamp", 
               user = "user_id") %>%
  funnel_step(moment_type = "registration",
              type = "first-firstafter") %>%
  funnel_step(moment_type = "purchase",
              type = "first-any")
#> # A tibble: 5 x 4
#>   user_id timestamp_landing timestamp_registration timestamp_purchase
#>     <dbl> <chr>             <chr>                  <chr>             
#> 1       3 2019-05-01        2019-06-01             2019-06-04        
#> 2       1 2019-07-01        2019-07-02             2019-07-07        
#> 3       1 2019-07-01        2019-07-02             2019-07-10        
#> 4       2 2019-08-01        2019-08-15             <NA>              
#> 5       4 2019-06-13        <NA>                   <NA>
```

If you use a `type` that allows multiple moments of one type for a user,
like “first-any”, you will get more rows per user rather than more
columns. For example, user 1 had two purchases, so she now has two rows.
The `timestamp_landing` and `timestamp_registration` is the same for
both rows, but they have a different `timestamp_purchase`.

Finally, you can use the `summarize_funnel()` to understand how many and
what percentage of people make it through to each next step of the
funnel. We can also switch to `funnel_steps()` to shorten our code a
bit: we give it a character vector of `moment_types` in order and the
`type` for each step.

``` r
activity %>%
  funnel_start(moment_type = "landing", 
               moment = "event", 
               tstamp = "timestamp", 
               user = "user_id") %>%
  funnel_steps(moment_types = c("registration", "purchase"),
              type = "first-firstafter") %>%
  summarize_funnel()
#> # A tibble: 3 x 4
#>   moment_type  nb_step pct_cumulative pct_step
#>   <fct>          <int>          <dbl>    <dbl>
#> 1 landing            4           1      NA    
#> 2 registration       3           0.75    0.75 
#> 3 purchase           2           0.5     0.667
```

`nb_step` is how many users made it to each step, `pct_cumulative` is
what percent that is out of the original step, and `pct_step` is what
percentage that is out of those who made it to the previous step. So in
our case, 2 people had a purchase, which is 50% of the people who landed
but 66% of those who registered.

## Reporting bugs and adding features

If you find any bugs or have a feature request or question, please
[create an issue](https://github.com/robinsones/funneljoin/issues/new).
If you’d like to add a feature, tests, or other functionality, please
also make an issue first and let’s discuss\!

Funneljoin was developed at DataCamp by Anthony Baker, David Robinson,
and Emily Robinson and continues to be maintained primarily by Emily.


NEWS.md
# funneljoin (development version)
* Allow users to join by multiple columns in `after_join()` with the `by_user` argument and in `funnel_start()` with the `user` argument. 

# funneljoin 0.1.0

* Initial release for time-based joins
* Added a `NEWS.md` file to track changes to the package.



R/after_join.R
#' Join tables based on one event happening after another
#'
#' Join two tables based on observations in one table happening after
#' observations in the other. Each table must have a user_id column,
#' which must always match for two observations to be joined,
#' and a time column, which must be greater in \code{y} than in \code{x} for
#' the two to be joined.
#' Supports all types of dplyr joins (inner, left, anti, etc.) and requires a
#' type argument to specify which observations in a funnel get kept
#' (see details for supported types).
#'
#' @param x A tbl representing the first event to occur in the funnel.
#' @param y A tbl representing an event to occur in the funnel.
#' @param by_time A character vector to specify the time columns in x and y.
#'   This would typically be a datetime or a date column. These columns are used to
#'   filter for time y being after time x.
#' @param by_user A character vector to specify the user or identity columns in
#'   x and y.
#' @param mode The method used to join: "inner", "full", "anti", "semi",
#'   "right", "left". Each also has its own function, such as
#'   \code{after_inner_join}.
#' @param type The type of funnel used to distinguish between event pairs, such
#'   as "first-first", "last-first", or "any-firstafter". See details for more.
#' @param max_gap Optional: the maximum gap allowed between events. Can be a
#'   integer representing the number of seconds or a difftime object, such as
#'   \code{as.difftime(2, units = "hours")}.
#' @param min_gap Optional: the maximum gap allowed between events. Can be a
#'   integer representing the number of seconds or a difftime object, such as
#'   \code{as.difftime(2, units = "hours")}.
#' @param gap_col Whether to include a numeric column, \code{.gap},
#'   with the time difference in seconds between the events.
#' @param suffix If there are non-joined duplicate variables in x and y,
#' these suffixes will be added to the output to disambiguate them.
#' Should be a character vector of length 2.
#' @importFrom magrittr %>%
#'
#' @details
#'
#' \code{type} can be any combination of \code{first}, \code{last}, \code{any}, \code{lastbefore}, \code{firstwithin} with \code{first}, \code{last}, \code{any},  \code{firstafter}. Some common ones you may use include:
#' \describe{
#'   \item{first-first}{Take the earliest x and y for each user \bold{before} joining. For example, you want the first time someone entered an experiment, followed by the first time someone \bold{ever} registered. If they registered, entered the experiment, and registered again, you do not want to include that person.}
#'   \item{first-firstafter}{Take the first x, then the first y after that. For example, you want when someone first entered an experiment and the first course they started afterwards. You don't care if they started courses before entering the experiment. }
#'   \item{lastbefore-firstafter}{First x that's followed by a y before the next x. For example, in last click paid ad attribution, you want the last time someone clicked an ad before the first subscription they did afterward.}
#'   \item{any-firstafter}{Take all Xs followed by the first Y after it. For example, you want all the times someone visited a homepage and their first product page they visited afterwards.}
#'   \item{any-any}{Take all Xs followed by all Ys. For example, you want all the times someone visited a homepage and \bold{all} the product pages they saw afterward.}
#'   }
#'
#' @export
#'
#' @examples
#'
#' library(dplyr)
#' landed <- tribble(
#'   ~user_id, ~timestamp,
#'   1, "2018-07-01",
#'   2, "2018-07-01",
#'   2, "2018-07-01",
#'   3, "2018-07-02",
#'   4, "2018-07-01",
#'   4, "2018-07-04",
#'   5, "2018-07-10",
#'   5, "2018-07-12",
#'   6, "2018-07-07",
#'   6, "2018-07-08"
#' ) %>%
#'   mutate(timestamp = as.Date(timestamp))
#'
#' registered <- tribble(
#'   ~user_id, ~timestamp,
#'   1, "2018-07-02",
#'   3, "2018-07-02",
#'   4, "2018-06-10",
#'   4, "2018-07-02",
#'   5, "2018-07-11",
#'   6, "2018-07-10",
#'   6, "2018-07-11",
#'   7, "2018-07-07"
#' ) %>%
#'  mutate(timestamp = as.Date(timestamp))
#'
#' after_inner_join(landed, registered, by_user = "user_id",
#'            by_time = "timestamp", type = "first-first")
#'
#' # You can use different methods of joining:
#' after_left_join(landed, registered, by_user = "user_id",
#'            by_time = "timestamp", type = "first-first")
#'
#' after_anti_join(landed, registered, by_user = "user_id",
#'            by_time = "timestamp", type = "any-any")
#'
after_join <- function(x,
                       y,
                       by_time,
                       by_user,
                       mode = "inner",
                       type = "first-first",
                       max_gap = NULL,
                       min_gap = NULL,
                       gap_col = FALSE,
                       suffix = c(".x", ".y")) {

  if (!is.null(attr(x, "after_join")) & inherits(x, "tbl_lazy")) {
    stop("You can not do multiple after joins in a row remotely. Please pull your data locally.")
  }

  types <- strsplit(type, '\\-')[[1]]

  if (length(types) != 2) {
    stop("type argument only supports pairs")
  }

  type_x <- match.arg(types[1], c("first", "last", "any", "lastbefore", "firstwithin"))
  type_y <- match.arg(types[2], c("first", "last", "any", "firstafter"))

  user_xy <- dplyr::common_by(by_user, x, y)

  if (length(by_user) > 1) {
    # combine user columns into one
    new_x_user_col_name <- paste(user_xy$x, collapse = "_")
    x <- x %>%
      tidyr::unite(!!dplyr::sym(new_x_user_col_name), !!dplyr::quo(user_xy$x))
    user_xy$x <- new_x_user_col_name

    new_y_user_col_name <- paste(user_xy$y, collapse = "_")
    y <- y %>%
      tidyr::unite(!!dplyr::sym(new_y_user_col_name), !!dplyr::quo(user_xy$y))
    user_xy$y <- new_y_user_col_name
  }

  if (length(by_time) > 1) {
    stop("Joining on multiple time columns is not supported. Check the by_time argument.")
  }

  time_xy <- dplyr::common_by(by_time, x, y)

  x_i <- x %>%
    dplyr::arrange(!!dplyr::sym(user_xy$x), !!dplyr::sym(time_xy$x)) %>%
    dplyr::mutate(..idx = row_number())

  y_i <- y %>%
    dplyr::arrange(!!dplyr::sym(user_xy$y), !!dplyr::sym(time_xy$y)) %>%
    dplyr::mutate(..idy = row_number())

  if (type_x %in% c("first", "last")) {
    x_i <- x_i %>%
      distinct_events(time_col = time_xy$x,
                      user_col = user_xy$x,
                      type = type_x)
  }

  if (type_y %in% c("first", "last")) {
    y_i <- y_i %>%
      distinct_events(time_col = time_xy$y,
                      user_col = user_xy$y,
                      type = type_y)
  }


  # Handle the case when columns with the same name are appended with .x & .y
  if (time_xy$x == time_xy$y || time_xy$x %in% colnames(y_i) || time_xy$y %in% colnames(x_i)) {
    time_xy <- list(x = paste0(time_xy$x, ".x"),
                    y = paste0(time_xy$y, ".y"))
  }

  # Get all the matching rows
  pairs <- x_i %>%
    dplyr::inner_join(y_i, by = user_xy) %>%
    dplyr::filter(!!dplyr::sym(time_xy$x) <= !!dplyr::sym(time_xy$y))

  if (!is.null(max_gap) || !is.null(min_gap)) {
    pairs <- filter_in_gap(pairs = pairs,
                           max_gap = max_gap,
                           min_gap = min_gap,
                           time_xy = time_xy,
                           user_xy = user_xy)
  }

  if (type_x == "firstwithin") {
    pairs <- pairs %>%
      distinct_events(time_col = time_xy$x,
                      user_col = "..idy",
                      type = "first")
    }

  if (type_y == "firstafter") {
    pairs <- pairs %>%
      distinct_events(time_col = time_xy$y,
                      user_col = "..idx",
                      type = "first")
  }

  if (type_x == "lastbefore") {
    pairs <- pairs %>%
      distinct_events(time_col = time_xy$x,
                      user_col = "..idy",
                      type = "last")
  }

  if (gap_col) {
    if (inherits(pairs, "tbl_lazy")) {
      time_difference <- dplyr::sql(glue::glue('DATEDIFF("seconds",
                                               "{ time_xy$x }",
                                               "{ time_xy$y }")::integer'))

      pairs <- pairs %>%
        dplyr::mutate(.gap = time_difference) %>%
        dplyr::select(..idx, ..idy, .gap)
    }
    else {
      pairs <- pairs %>%
        dplyr::mutate(.gap = as.numeric(difftime(!!dplyr::sym(time_xy$y),
                                                 !!dplyr::sym(time_xy$x), units = "secs"))) %>%
        dplyr::select(..idx, ..idy, .gap)
    }
  } else {
    pairs <- pairs %>%
      dplyr::select(..idx, ..idy)
  }

  join_func <- switch(mode,
                      inner = dplyr::inner_join,
                      left = dplyr::left_join,
                      right = dplyr::right_join,
                      full = dplyr::full_join,
                      semi = dplyr::semi_join,
                      anti = dplyr::anti_join
  )

  if (is.null(join_func)) {
    stop("Unknown joining mode: ", mode)
  }

    if (mode %in% c("inner", "left", "right", "full")) {

    if (length(by_user) > 1) {
      by_user_new <- c(new_x_user_col_name, new_y_user_col_name)
      ret <- x_i %>%
        join_func(pairs, by = "..idx") %>%
        join_func(y_i, by = c(by_user_new, "..idy" = "..idy"), suffix = suffix) %>%
        dplyr::select(-..idx, -..idy)
    }
    else {
      ret <- x_i %>%
        join_func(pairs, by = "..idx") %>%
        join_func(y_i, by = c(by_user, "..idy" = "..idy"), suffix = suffix) %>%
        dplyr::select(-..idx, -..idy)
    }
  } else if (mode %in% c("semi", "anti")) {
    ret <- x_i %>%
      join_func(pairs, by = "..idx") %>%
      dplyr::select(-..idx)
  }

  attr(ret, "after_join") <- 1
  ret
}

#' @rdname after_join
#' @export
after_inner_join <- function(x, y, by_time, by_user, type, max_gap = NULL,  min_gap = NULL, gap_col = FALSE, suffix = c(".x", ".y")) {
  after_join(x, y, by_time, by_user,
             mode = "inner", type = type, max_gap = max_gap, min_gap = min_gap, gap_col = gap_col, suffix = suffix)
}

#' @rdname after_join
#' @export
after_left_join <- function(x, y, by_time, by_user, type, max_gap = NULL, min_gap = NULL, gap_col = FALSE, suffix = c(".x", ".y")) {
  after_join(x, y, by_time, by_user,
             mode = "left", type = type, max_gap = max_gap, min_gap = min_gap, gap_col = gap_col, suffix = suffix)
}

#' @rdname after_join
#' @export
after_right_join <- function(x, y, by_time, by_user, type, max_gap = NULL, min_gap = NULL, gap_col = FALSE, suffix = c(".x", ".y")) {
  after_join(x, y, by_time, by_user,
             mode = "right", type = type, max_gap = max_gap, min_gap = min_gap, gap_col = gap_col, suffix = suffix)
}

#' @rdname after_join
#' @export
after_full_join <- function(x, y, by_time, by_user, type, max_gap = NULL, min_gap = NULL, gap_col = FALSE, suffix = c(".x", ".y")) {
  after_join(x, y, by_time, by_user,
             mode = "full", type = type, max_gap = max_gap, min_gap = min_gap, gap_col = gap_col, suffix = suffix)
}

#' @rdname after_join
#' @export
after_anti_join <- function(x, y, by_time, by_user, type, max_gap = NULL, min_gap = NULL, gap_col = FALSE, suffix = c(".x", ".y")) {
  after_join(x, y, by_time, by_user,
             mode = "anti", type = type, max_gap = max_gap, min_gap = min_gap, gap_col = gap_col, suffix = suffix)
}

#' @rdname after_join
#' @export
after_semi_join <- function(x, y, by_time, by_user, type, max_gap = NULL, min_gap = NULL,
                            gap_col = FALSE, suffix = c(".x", ".y")) {
  after_join(x, y, by_time, by_user,
             mode = "semi", type = type, max_gap = max_gap, min_gap = min_gap, gap_col = gap_col, suffix = suffix)
}
